app/src/main/java/com/github/polybooks/FillSaleActivity.kt

```diff
@@ -2,63 +2,209 @@ package com.github.polybooks
+import com.github.polybooks.utils.UIManip.enableButton
+import com.github.polybooks.utils.url2json
+import com.google.firebase.Timestamp
+import java.lang.Exception
+import java.text.DateFormat
 import java.util.concurrent.CompletableFuture
 
+
 /**
  * This activity receives the ISBN, either manually inputted from AddSale or deduced from the scanned barcode,
  * shows the retrieved data, but does not allow modification of it, only confirmation,
  * and offers some additional manual fields such as price, condition, etc.
  */
-class FillSaleActivity : AppCompatActivity() {
+class FillSaleActivity : AppCompatActivity(), AdapterView.OnItemSelectedListener {
+
+    // TODO I would imagine that in the future, the dbs are global constants, but while writing this class, I'll instantiate one locally
+    private val salesDB = SaleDatabase()
+    private val bookDB = OLBookDatabase { string -> url2json(string) }
+
+    private lateinit var bookFuture: CompletableFuture<Book?>
+    private var bookConditionSelected: BookCondition? = null
+    private val dateFormat: DateFormat = DateFormat.getDateInstance(DateFormat.LONG)
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_fill_sale_fancy)
 
         // Get the Intent that started this activity and extract the string
         val stringISBN = intent.getStringExtra(ISBN)
-        // TODO for testing purpose, the ISBN will temporarily be displayed in the publisher field
-        findViewById<TextView>(R.id.filled_publisher)         .apply { text = stringISBN }
-
-        // Check if ISBN in our database: (could check ISBN validity before)
 
-        // TODO Commenting out the whole chunk as it depends on the API which is not ready yet.
-        /*
-        val book: CompletableFuture<Book> = CompletableFuture().supplyAsync {
-
-            bookDatabase.getBook(stringISBN)
-        }.handle { (book, err) ->
-            {
-                if (err == null) {
-        // Yes: Retrieve from our database data about the book
-                    findViewById<TextView>(R.id.filled_authors)         .apply { text = book.authors }
-                    findViewById<TextView>(R.id.filled_title)           .apply { text = book.title }
-                    findViewById<TextView>(R.id.filled_edition)         .apply { text = book.edition }
-                    findViewById<TextView>(R.id.filled_language)        .apply { text = book.language }
-                    findViewById<TextView>(R.id.filled_publisher)       .apply { text = book.publisher }
-                    findViewById<TextView>(R.id.filled_publish_date)    .apply { text = book.publishDate }
-                    findViewById<TextView>(R.id.filled_format)          .apply { text = book.format }
+        // Retrieve book data and display it if possible, else redirect with error toast
+        if(!stringISBN.isNullOrEmpty() && isbnHasCorrectFormat(stringISBN)) {
```

Comment:
I think it's exactly what I did ^^ regulariseISBN exists no more and has become StringUtils.isbnHasCorrectFormat

Reply:
Sorry, my bad, I did the reverse actually: I extracted regulariseISBN logic to StringUtils.isbnHasCorrectFormat and made regulariseISBN use isbnHasCorrectFormat.
I think it's fine like that, because isbnHasCorrectFormat returns a boolean while regulariseISBN transforms the string.

Reply:
Not exactly, ```isbnHasCorrectFormat()``` only accepts a subset of valid ISBN formats. What I did with ```regulariseISBN``` was to transform all ISBNs into a more uniform format and also correct some small formatting errors.
The UI should be lenient. The user could insert extra spaces or dashes in the ISBN.

Reply:
Hum, allowing the UI to be lenient is a good point because it's true that ISBN are sometimes written with spaces or dashes.
However constraining the UI is another decent alternative in my opinion:
If we allowed the user to type in letter, space, dashes, then yeah, we should be lenient and clean them up instead of forbidding them.
But if the user can only input numbers and every other option is not available, I think it's quite obvious to the user to input the isbn as an "int". And it's what the UI currently offers thanks to the number field type (instead of string field).
We could gather other opinions, I'm open to both options.